### PR TITLE
geos: 3.9.1 -> 3.10.2

### PR DESCRIPTION
**Pull request by willcohen**
###### Things done

Version bump of `geos` from 3.9.1 to 3.10.2, which also requires the use of cmake.

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kirillrdy** [comment]
I ventured on this path as well, and ran into issues

`geos-config` would print wrong paths for cflags and ldflags

I patched it by 

```
 postPatch = ''
    substituteInPlace tools/geos-config.in \
      --replace "@libdir@"         "@prefix@/lib" \
      --replace "@includedir@"         "@prefix@/include" \
  '';
```

seems to be working fine after that

also, you can probably remove `configureFlags`

**Comment by willcohen** [comment]
`nixpkgs-review` is getting stuck on hung up on python3.9-OpenSfM, but everything except for python-astropy seems to be generally working, with these changes:

```
$ nixpkgs-review pr 165322 --post-result
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/165322/head:refs/nixpkgs-review/1
From https://github.com/NixOS/nixpkgs
   2d09f87a891..855e992a7af  master                -> refs/nixpkgs-review/0
 + 5459b21241f...bf678f2ac0a refs/pull/165322/head -> refs/nixpkgs-review/1  (forced update)
$ git worktree add /home/wcohen/.cache/nixpkgs-review/pr-165322-4/nixpkgs 855e992a7af8461486f81906fca2a9ee1956c4cf
Preparing worktree (detached HEAD 855e992a7af)
Updating files: 100% (30077/30077), done.
HEAD is now at 855e992a7af Merge pull request #166251 from mweinelt/home-assistant
$ nix-env --option system x86_64-linux -f /home/wcohen/.cache/nixpkgs-review/pr-165322-4/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit --no-ff bf678f2ac0aeaa0f69d671d1d7783b6675434b50
Automatic merge went well; stopped before committing as requested
$ nix-env --option system x86_64-linux -f /home/wcohen/.cache/nixpkgs-review/pr-165322-4/nixpkgs -qaP --xml --out-path --show-trace --meta
1 package added:
geos39 (init at 3.9.2)

91 packages updated:
cloudcompare-unstable cura flatcam gdal gdal gdal gdal geos (3.9.1 → 3.10.2) gmt gplates grass libLAS librttopo libspatialite mapcache mapnik MapProxy mapserver merkaartor mod_tile-unstable moonraker-unstable OpenOrienteering-Mapper openspace osmscout-server paraview pdal perl5.32.1-Tirex perl5.34.0-Tirex postgis postgis postgis postgis python3.10-aplpy python3.10-basemap python3.10-cartopy python3.10-descartes python3.10-fiona python3.10-flask-admin python3.10-GeoAlchemy2 python3.10-geopandas python3.10-imgaug python3.10-OpenSfM-unstable python3.10-osmnx python3.10-OWSLib python3.10-preprocess-cancellation python3.10-pygeos python3.10-pygmt python3.10-pyosmium python3.10-pyproj python3.10-python-mapnik-unstable python3.10-rasterio python3.10-Shapely python3.10-spectral-cube python3.10-uranium python3.10-worldengine python3.9-aplpy python3.9-basemap python3.9-cartopy python3.9-descartes python3.9-fiona python3.9-flask-admin python3.9-GeoAlchemy2 python3.9-geopandas python3.9-imgaug python3.9-mask-rcnn python3.9-OpenSfM-unstable python3.9-osmnx python3.9-OWSLib python3.9-preprocess-cancellation python3.9-pygeos python3.9-pygmt python3.9-pyosmium python3.9-pyproj python3.9-python-mapnik-unstable python3.9-rasterio python3.9-Shapely python3.9-spectral-cube python3.9-uranium python3.9-worldengine pytrainer qgis qgis qlandkartegt qmapshack readosm saga spatialite-tools sumo t-rex udig valhalla

$ nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/wcohen/.cache/nixpkgs-review/pr-165322-4/build.nix
error: builder for '/nix/store/h5c6lhkhjv7lwblfylbmgsmh64sk1snf-python3.10-aplpy-2.0.3.drv' failed with exit code 4;
       last 10 log lines:
       > Executing pythonRemoveTestsDir
       > Finished executing pythonRemoveTestsDir
       > running install tests
       > Matplotlib created a temporary config/cache directory at /build/matplotlib-e2elue96 because the default path (/homeless-shelter/.config/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
       > ImportError while loading conftest '/build/APLpy-2.0.3/aplpy/conftest.py'.
       > aplpy/__init__.py:14: in <module>
       >     from .core import FITSFigure  # noqa
       > aplpy/core.py:18: in <module>
       >     from astropy.nddata.utils import block_reduce
       > E   ImportError: cannot import name 'block_reduce' from 'astropy.nddata.utils' (/nix/store/3s6bfl12l6d9f976yz631b55hvqmg39f-python3.10-astropy-5.0/lib/python3.10/site-packages/astropy/nddata/utils.py)
       For full logs, run 'nix log /nix/store/h5c6lhkhjv7lwblfylbmgsmh64sk1snf-python3.10-aplpy-2.0.3.drv'.
error: 1 dependencies of derivation '/nix/store/pszg1qj10bqcpm3z1biaqbh72amfnrzj-python3.10-spectral-cube-0.6.0.drv' failed to build
error: builder for '/nix/store/2xnxavh0kc82xhdph0mccfi6nq8g9l34-python3.9-aplpy-2.0.3.drv' failed with exit code 4;
       last 10 log lines:
       > Executing pythonRemoveTestsDir
       > Finished executing pythonRemoveTestsDir
       > running install tests
       > Matplotlib created a temporary config/cache directory at /build/matplotlib-q3xlbe5m because the default path (/homeless-shelter/.config/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
       > ImportError while loading conftest '/build/APLpy-2.0.3/aplpy/conftest.py'.
       > aplpy/__init__.py:14: in <module>
       >     from .core import FITSFigure  # noqa
       > aplpy/core.py:18: in <module>
       >     from astropy.nddata.utils import block_reduce
       > E   ImportError: cannot import name 'block_reduce' from 'astropy.nddata.utils' (/nix/store/lf1kcprqm147jym5fykkgfpg6ddv4ici-python3.9-astropy-5.0/lib/python3.9/site-packages/astropy/nddata/utils.py)
       For full logs, run 'nix log /nix/store/2xnxavh0kc82xhdph0mccfi6nq8g9l34-python3.9-aplpy-2.0.3.drv'.
error: 1 dependencies of derivation '/nix/store/96ya0jmqzriq3v7ihkc6v0fr27mqjwdx-python3.9-spectral-cube-0.6.0.drv' failed to build
[2/55/59 built (2 failed), 6 copied (43.2/43.2 MiB), 17.8 MiB DL] building python3.9-OpenSfM-unstable-2022-03-10 (pytestCheckPhase): opensfm/test/test_reconstruction_alignment.py ........                   [ 75%]
```

File: pkgs/development/libraries/geos/default.nix [hunk 1]
```diff
@@ -1,22 +1,33 @@
...
 
+  buildInputs = [ cmake ];
+
   enableParallelBuilding = true;
 
-  # https://trac.osgeo.org/geos/ticket/993
-  configureFlags = lib.optional stdenv.isAarch32 "--disable-inline";
+  postPatch = ''
+    substituteInPlace tools/geos-config.in \
+      --replace "@libdir@" "@prefix@/lib" \
```
Comment (kirillrdy) [inline pkgs/development/libraries/geos/default.nix:21]: maybe its still worth investigating what actually sets `@libdir@` and `@includedir@` to wrong values
maybe that will fix the other issues `valhalla`
if i get some free time today I'll have a look at `valhalla`, but don't want to block this from getting merged

Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix:21]: I got as far as noting that valhalla is getting stuck with 3.10.x for brew users with https://github.com/valhalla/valhalla/issues/3388, but it doesn't seem totally resolved with them either, leading me to figure the best case was to just leave valhalla with 3.9 and let the rest move forward for now.

Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix:21]: If, as noted in https://github.com/valhalla/valhalla/issues/3388 and https://github.com/libgeos/geos/issues/500, then https://github.com/libgeos/geos/issues/543 will be reflected with 3.11.0, so I'd consider this the best resolution for 3.10, and geos39 can be removed when geos bumps to 3.11.

File: pkgs/development/libraries/geos/default.nix [hunk 2]
```diff
@@ -1,22 +1,33 @@
...
-    sha256 = "sha256-fmMFB9ysncB1ZdJJom8GoVyfWwxS3SkSmg49OB1+OCo=";
+    sha256 = "sha256-ULvFmaw4a0wrOWLcxBHwBAph8gSq7066ciXs3Qz0VxU=";
   };
 
+  buildInputs = [ cmake ];
+
   enableParallelBuilding = true;
 
-  # https://trac.osgeo.org/geos/ticket/993
-  configureFlags = lib.optional stdenv.isAarch32 "--disable-inline";
```
Comment (kirillrdy) [inline pkgs/development/libraries/geos/default.nix:15]: question: for `geos39` do we want to keep this line ?

Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix:15]: Good catch. Per https://github.com/libgeos/geos/issues/512 I think we should leave it in for now.

Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix:15]: Updated with the latest force push.

File: pkgs/development/libraries/geos/default.nix [hunk 3]
```diff
@@ -1,22 +1,33 @@
...
-  version = "3.9.1";
+  version = "3.10.2";
 
   src = fetchurl {
     url = "https://download.osgeo.org/geos/${pname}-${version}.tar.bz2";
-    sha256 = "sha256-fmMFB9ysncB1ZdJJom8GoVyfWwxS3SkSmg49OB1+OCo=";
+    sha256 = "sha256-ULvFmaw4a0wrOWLcxBHwBAph8gSq7066ciXs3Qz0VxU=";
   };
 
+  buildInputs = [ cmake ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/geos/default.nix:15]:
```suggestion
  nativeBuildInputs = [ cmake ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Review by kirillrdy — APPROVED** [review]
I've also ran nix-review on `x86_64-linux`

```
nix --experimental-features nix-command build --no-link --keep-going --option build-use-sandbox relaxed -f /home/kirillvr/.cache/nixpkgs-review/pr-165322-3/build.nix
error: builder for '/nix/store/h5c6lhkhjv7lwblfylbmgsmh64sk1snf-python3.10-aplpy-2.0.3.drv' failed with exit code 4;
       last 10 log lines:
       > Executing pythonRemoveTestsDir
       > Finished executing pythonRemoveTestsDir
       > running install tests
       > Matplotlib created a temporary config/cache directory at /build/matplotlib-3t7gk_px because the default path (/homeless-shelter/.config/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
       > ImportError while loading conftest '/build/APLpy-2.0.3/aplpy/conftest.py'.
       > aplpy/__init__.py:14: in <module>
       >     from .core import FITSFigure  # noqa
       > aplpy/core.py:18: in <module>
       >     from astropy.nddata.utils import block_reduce
       > E   ImportError: cannot import name 'block_reduce' from 'astropy.nddata.utils' (/nix/store/3s6bfl12l6d9f976yz631b55hvqmg39f-python3.10-astropy-5.0/lib/python3.10/site-packages/astropy/nddata/utils.py)
       For full logs, run 'nix log /nix/store/h5c6lhkhjv7lwblfylbmgsmh64sk1snf-python3.10-aplpy-2.0.3.drv'.
error: 1 dependencies of derivation '/nix/store/pszg1qj10bqcpm3z1biaqbh72amfnrzj-python3.10-spectral-cube-0.6.0.drv' failed to build
error: builder for '/nix/store/2xnxavh0kc82xhdph0mccfi6nq8g9l34-python3.9-aplpy-2.0.3.drv' failed with exit code 4;
       last 10 log lines:
       > Executing pythonRemoveTestsDir
       > Finished executing pythonRemoveTestsDir
       > running install tests
       > Matplotlib created a temporary config/cache directory at /build/matplotlib-23ob226k because the default path (/homeless-shelter/.config/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
       > ImportError while loading conftest '/build/APLpy-2.0.3/aplpy/conftest.py'.
       > aplpy/__init__.py:14: in <module>
       >     from .core import FITSFigure  # noqa
       > aplpy/core.py:18: in <module>
       >     from astropy.nddata.utils import block_reduce
       > E   ImportError: cannot import name 'block_reduce' from 'astropy.nddata.utils' (/nix/store/lf1kcprqm147jym5fykkgfpg6ddv4ici-python3.9-astropy-5.0/lib/python3.9/site-packages/astropy/nddata/utils.py)
       For full logs, run 'nix log /nix/store/2xnxavh0kc82xhdph0mccfi6nq8g9l34-python3.9-aplpy-2.0.3.drv'.
error: 1 dependencies of derivation '/nix/store/96ya0jmqzriq3v7ihkc6v0fr27mqjwdx-python3.9-spectral-cube-0.6.0.drv' failed to build
error: 4 dependencies of derivation '/nix/store/0v981ccfa0c5ww344cb311v3yi5vb5h6-env.drv' failed to build
error: 1 dependencies of derivation '/nix/store/zxzq2x1kdzr3fyj7j3xcw3541b64jxjr-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/165322

5 packages marked as broken and skipped:
openspace python310Packages.basemap python310Packages.descartes python39Packages.basemap python39Packages.descartes

8 packages failed to build:
apacheHttpdPackages.mod_tile mapnik perl532Packages.Tirex perl534Packages.Tirex python310Packages.aplpy python310Packages.spectral-cube python39Packages.aplpy python39Packages.spectral-cube

79 packages built:
cloudcompare cura flatcam gdal gdal_2 geos geos39 gmt gplates grass haskellPackages.hgeos libLAS librttopo libspatialite mapcache mapproxy mapserver merkaartor moonraker openorienteering-mapper osmscout-server paraview pdal postgresql11Packages.postgis postgresql12Packages.postgis postgresqlPackages.postgis postgresql14Packages.postgis python310Packages.cartopy python310Packages.fiona python310Packages.flask-admin python310Packages.gdal python310Packages.geoalchemy2 python310Packages.geopandas python310Packages.imgaug python310Packages.opensfm python310Packages.osmnx python310Packages.owslib python310Packages.preprocess-cancellation python310Packages.pygeos python310Packages.pygmt python310Packages.pyosmium python310Packages.pyproj python310Packages.python-mapnik python310Packages.rasterio python310Packages.shapely python310Packages.uranium python310Packages.worldengine python39Packages.cartopy python39Packages.fiona python39Packages.flask-admin python39Packages.geoalchemy2 python39Packages.geopandas python39Packages.imgaug python39Packages.mask-rcnn python39Packages.opensfm python39Packages.osmnx python39Packages.owslib python39Packages.preprocess-cancellation python39Packages.pygeos python39Packages.pygmt python39Packages.pyosmium python39Packages.pyproj python39Packages.python-mapnik python39Packages.rasterio python39Packages.shapely python39Packages.uranium worldengine-cli pytrainer qgis qgis-ltr qlandkartegt qmapshack readosm saga spatialite_tools sumo t-rex udig valhalla

error: build log of '/nix/store/q57677cl70s9izl9sc4ncwmyhya3b7l6-mod_tile-unstable-2017-01-08.drv' is not available
error: build log of '/nix/store/693q5w66lfavla3aixrmgf0licp3xr7x-mod_tile-unstable-2017-01-08' is not available
```

File: pkgs/development/libraries/geos/default.nix (hunk 3 above)
Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix]: Change made with latest force push!

File: pkgs/development/libraries/geos/default.nix [hunk 4]
```diff
@@ -1,22 +1,33 @@
...
 
   src = fetchurl {
     url = "https://download.osgeo.org/geos/${pname}-${version}.tar.bz2";
-    sha256 = "sha256-fmMFB9ysncB1ZdJJom8GoVyfWwxS3SkSmg49OB1+OCo=";
+    sha256 = "sha256-ULvFmaw4a0wrOWLcxBHwBAph8gSq7066ciXs3Qz0VxU=";
   };
 
+  nativeBuildInputs = [ cmake ];
+
   enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/geos/default.nix:18]:
```suggestion
```
already set by cmake

Comment (willcohen) [reply pkgs/development/libraries/geos/default.nix]: Change made with latest force push. Thanks for catching!

**Comment by kirillrdy** [comment]
potentially fixing python3.pkgs.aplpy build issue https://github.com/NixOS/nixpkgs/pull/166789

**Pull request merged**